### DISCUSSION TITLE
add cheats (add guide for cmake)

### DISCUSSION THREAD

**Discussion by fheroes2bugs** [ihhub/fheroes2#7843] [General]
### Discussed in https://github.com/ihhub/fheroes2/discussions/5265

<div type='discussions-op-text'>

<sup>Originally posted by **JurajHanakovic** April 11, 2022</sup>
This game/sourceport/engine-recreation really needs to add cheats.

Some scenarios are extra hard for casual and non-casual player alike. Especially when playing in non-optimal ways.
And I don't want to replay 2h+ scenario if I don't play optimaly all the time.
You know, games are supposed to be fun, not a chore to play.

I'm quite certain that many other players will appreciate cheat options to be added.</div>

I have seen a lot of messages from contributors "just simply use debug mode in cmake", but for not programmers, it is not "simply".

 You have to google guides about cmake, then you download it, press configure (the guide from youtube says!) and get error "error in configuration process, project files may be invalid". Ok, looks like downloading cmake was the wrong way, we  download vscode, add cmake tools extensions (already a lot of videos on youtube watched how to do it), and press ctrl+shift+p and get error "configure failed". Ok, seems like vscode guide from youtube was wrong and we just wasted time. Ok, downloading visual studio with cmake extensions, click on cmakelists.txt and getting the error "Could not find a package configuration file provided by "SDL2" with any of  the following names: SDL2Config.cmake    sdl2-config.cmake". Tired of googling and installing software only way is to delete it all. For programmers is easy, for other is not. I am sure, after reading "just simply use debug mode in cmake" 1 in a 100 people will go and actually build it in debug mode, others will just close the window and forget about it, not wasting their time, like I did.

So, if it so easy, can someone make a guide? Like:
1. download programm_name.exe
2. install it
3. click here and here and it just works.

Or doing a video, it will be like 40 second duration, but it will help a lot of people.

**Comment by ihhub** [comment]
@fheroes2bugs ,

1. Please read 2 **whole** discussions about cheats. There are reasons why it wasn't implemented. Nobody proposed a valid solution for highlighted problems.
2. Which difficulty level you chose to play the map and failed?
3. You are using Windows. VSCode is never mentioned by us and you cannot compile with it. Please read the official development instructions and not to blame developers for misguidance.

**Reply by fheroes2bugs** [reply]
@ihhub 
1. I have read #5265 #7635 #3824 and a few discuccions on reddit, soundforge and few heroes-forums - all about why not enable cheats in fheroes. I don't discuss if it is needed in game or not. But in all that discussions same (looks so) contributor keeps answering "build in debug mode in cmake and you can enable cheats" and then the discission is ended, no one is answering "thank you I succeed". It looks like none is even trying to build, they just close the window.
2. Useful to analyze AI, not to cheat to beat computer.
3. As far as I know, VSCode works perfect on Windows too, but that is not the question, I could do it with visual studio too, but again got error. I know, that I am doing something wrong. And I am the first person that asks "how exactly do I need to build game in debug mode in cmake".
Where can I find "official development instructions"? How many people will read it? The issue is about just to add small guide or video, so people wouldn't need to google and read the "official development instructions".

**Reply by oleg-derevenetz** [reply]
> VSCode works perfect on Windows too

VSCode is just an editor with quirks. It requires external plugins, toolsets and compiler to deal with a C++ project. All this needs to be installed separately. MSVC includes all the necessary components right from scratch. That's why we recommend to use MSVC on Windows.

> Where can I find "official development instructions"?

They are right there, in the main README file. "To build the project from source, please follow this guide" with the link on "this guide" in the "Contribution and Development" section.

> The issue is about just to add small guide or video

If a person has not bothered to read the main README of a small size and click on the link, how will the video help here? What would be the difference if this link was to a video? You would have missed it the same way.

**Reply by ihhub** [reply]
@fheroes2bugs , for AI analysis debug build is exactly the best tool to use, it even has AI control for human players. So there is no need for cheats in the AI tests.

**Reply by zenseii** [reply]
I'm also very against a video guide because with projects like these they get quickly outdated, which leads to users later coming and asking why they can't build when they followed a video guide from 5 years ago.

Setting up the build environment for fheroes2 is not something we recommend for people who just want to play the game, and therefore the guide is not made to be easy to do for those people. Instead the guide is made for people who want to contribute or to tinker with the code and engine, and it does not require prior programming experience but it is highly recommended. However, it does require the ability and patience to read the guide and to be able to follow it, and if you still run into problems, be able to ask for help.

I don't understand where you got the idea about cmake being the only possibility. The guide clearly says you can, and it is highly recommended to, use visual studio (for those without much prior experience), and yes you do have to be able to understand that "visual studio" is not the same thing as "visual code" - just like how winzip is not the same thing as winrar.

**Reply by fheroes2bugs** [reply]
So this is the "readme": https://github.com/ihhub/fheroes2/blob/master/docs/README_cmake.md

![image](https://github.com/ihhub/fheroes2/assets/145407420/95a84ac3-4003-441f-be6a-e5f4caff842b)
 Here for debug mode I just change from "release" to "debug". Then I go to C:\Users\user_name\Desktop\fheroes2-1.0.8\fheroes2-1.0.8\build\Debug 
![image](https://github.com/ihhub/fheroes2/assets/145407420/d9d3b714-44cd-4fc9-a773-2bc011554332)


...and copy the files to my fheroes directory. What to do next? Where can I "ebable cheats"? Open the map, for example?

**Reply by zenseii** [reply]
Hi, @fheroes2bugs. This guide is better suited for what you want to do in my opinion:
https://github.com/ihhub/fheroes2/blob/master/docs/DEVELOPMENT.md

Visual Studio Community edition is free to use.

**Comment by fheroes2bugs** [comment]
> Hi, @fheroes2bugs. This guide is better suited for what you want to do in my opinion: https://github.com/ihhub/fheroes2/blob/master/docs/DEVELOPMENT.md
> 
> Visual Studio Community edition is free to use.

Well, after that, next, where can I "ebable cheats"? Open the map, for example?

**Reply by zenseii** [reply]
There is not really a concept of cheats when you develop. With the skills you can quickly change the code to immediately win a map for example, but that's not a "cheat".

Make sure you are building a "debug" build and not a "release" build. Find fheroes2.cfg. Change to the debug value there in one of the parameters. You can find the debug value in logging.h. I don't remember what it is.

As any aspiring programmer your best tool is your curiosity to try out things and figure out what they do, but making sure you know how to go back on the changes you make.

**Reply by fheroes2bugs** [reply]
I when I build, I made a "debug" config, but can't see any "fheroes2.cfg".

![image](https://github.com/ihhub/fheroes2/assets/145407420/83988c6b-7b3d-46a5-b550-b196394cb970)

Only "timidity.cfg", but that is not what is needed.
It should be "fheroes2.cfg"?
A lot of files with "fheroes", but none is cfg:
![image](https://github.com/ihhub/fheroes2/assets/145407420/2cfe4631-7cb6-4aac-85af-0c8f914df343)

**Reply by zenseii** [reply]
fheroes2.cfg is found in c:\users\yourUserName\AppData\Roaming\fheroes2\

**Reply by fheroes2bugs** [reply]
So i should change 5460 to something else? Can't see values in logging.h

![image](https://github.com/ihhub/fheroes2/assets/145407420/9ccbcffa-1a44-467b-8015-1f1ca8e78652)

**Reply by zenseii** [reply]
The values are right there in front of you. They are all the ones after the = operators. I believe you have to convert them to decimal from hexadecimal, but you can try to just copy them with the 0x part. Not sure which of them is the right one.

**Reply by fheroes2bugs** [reply]
Thank you, finally, I succeed.
Tried one by one these values, and for open map it is "DBG_DEVEL"

![image](https://github.com/ihhub/fheroes2/assets/145407420/c7e5f9ef-5517-44c0-a416-14cb736d8675)


![image](https://github.com/ihhub/fheroes2/assets/145407420/9bd07650-e2c3-4618-a020-83aa21678c87)

**Comment by sroziewski** [comment]
I have found another solution for cheating, just change the source code, e.g. find an artifact on the map and change its behaviour.

**Reply by sroziewski** [reply]
![image](https://github.com/ihhub/fheroes2/assets/942525/38dba07c-7b4e-42a7-a70c-d4bc5101e71d)

**Reply by ihhub** [reply]
@sroziewski , it is obvious since this is an open source project. This is also not cheating but modding.
